session/session.go

```diff
@@ -0,0 +1,135 @@
+package session
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"launchpad.net/gozk"
+)
+
+type ZKSessionEvent uint
+
+var ErrZKSessionNotConnected = errors.New("unable to connect to ZooKeeper")
+
+const (
+	SessionClosed ZKSessionEvent = iota
+	SessionDisconnected
+	SessionReconnected
+	SessionFailed
+)
+
+type ZKSession struct {
+	servers     string
+	recvTimeout time.Duration
+	conn        *zookeeper.Conn
+	clientID    *zookeeper.ClientId
+	events      <-chan zookeeper.Event
+	sEvents     chan ZKSessionEvent
+	mu          sync.Mutex
+}
+
+func NewZKSession(servers string, recvTimeout time.Duration) (*ZKSession, <-chan ZKSessionEvent, error) {
+	conn, events, err := zookeeper.Dial(servers, recvTimeout)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	s := &ZKSession{
+		servers:     servers,
+		recvTimeout: recvTimeout,
+		conn:        conn,
+		clientID:    conn.ClientId(),
+		events:      events,
+		sEvents:     make(chan ZKSessionEvent),
+	}
+
+	event := <-events
+	if event.State != zookeeper.STATE_CONNECTED {
```

Comment:
dunno if it's better or worse, but you could write this as

```
if e := <-events; e.State != zookeeper.STATE_CONNECTED {
```

or even probably:
```
if (<-events).State != zookeeper.STATE_CONNECTED {
```